---
title: "ST_deconvolution"
output: html_document
date: "2024-05-28"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load dependencies

```{r load libraries}
suppressMessages(library(Seurat))
suppressMessages(library(SeuratDisk))
suppressMessages(library(schard))
suppressMessages(library(tidyverse))
suppressMessages(library(dplyr))
suppressMessages(library(stringr))
suppressMessages(library(ggplot2))
suppressMessages(library(knitr))
suppressMessages(library(corrplot))
suppressMessages(library(CARD))
suppressMessages(library(SeuratDisk))
#suppressMessages(library(Giotto))
```

Use scRNAseq ref data from Wu et al 2021 to deconvolute ST data 

# Read spaceranger output and create a seurat object

```{r create seurat object}

# Load 10x spaceranger output
bcsa <- Load10X_Spatial(data.dir = "~/Desktop/CIIR/data/visium_spatial_out/V10F24-114_A1_BCSA2_TumD3/outs/",
                        slice = "BCSA2TumD3")

# Calculate percentage of ribosomal genes
ribo <- grep("^RP[SL]", rownames(bcsa), value = T)
bcsa[["percent_ribo"]] <- PercentageFeatureSet(bcsa, features = ribo)

# Calculate percentage of mitochondrial genes
bcsa[["percent_mito"]] <- PercentageFeatureSet(object = bcsa, pattern = "^MT-")

# Calculate percentage of hemoglobin genes
hb <- grep("^HB", rownames(bcsa), value = T)
hb <- hb[-which(hb %in% c("HBEGF", "HBS1L", "HBP1"))]
bcsa[["percent_hb"]] <- PercentageFeatureSet(bcsa, features = hb)

# Remove mitochondria genes
bcsa[["Spatial"]]$counts <- bcsa[["Spatial"]]$counts[-grep("^MT-", rownames(bcsa[["Spatial"]]$counts)),]

# Save the seurat object
save(bcsa, file = "~/Desktop/CIIR/data/PAM50Bulk/results/BCSA2TumD3.RData")

############## BCSA2_TumD4

# Load 10x spaceranger output
bcsa <- Load10X_Spatial(data.dir = "~/Desktop/CIIR/data/visium_spatial_out/V10F24-114_B1_BCSA2_TumD4/outs/",
                        slice = "BCSA2TumD4")

# Calculate percentage of ribosomal genes
ribo <- grep("^RP[SL]", rownames(bcsa), value = T)
bcsa[["percent_ribo"]] <- PercentageFeatureSet(bcsa, features = ribo)

# Calculate percentage of mitochondrial genes
bcsa[["percent_mito"]] <- PercentageFeatureSet(object = bcsa, pattern = "^MT-")

# Calculate percentage of hemoglobin genes
hb <- grep("^HB", rownames(bcsa), value = T)
hb <- hb[-which(hb %in% c("HBEGF", "HBS1L", "HBP1"))]
bcsa[["percent_hb"]] <- PercentageFeatureSet(bcsa, features = hb)

# Remove mitochondria genes
bcsa[["Spatial"]]$counts <- bcsa[["Spatial"]]$counts[-grep("^MT-", rownames(bcsa[["Spatial"]]$counts)),]

# Save the seurat object
save(bcsa, file = "~/Desktop/CIIR/data/PAM50Bulk/results/BCSA2TumD4.RData")

############## BCSA4_TumA1

# Load 10x spaceranger output
bcsa <- Load10X_Spatial(data.dir = "~/Desktop/CIIR/data/visium_spatial_out/V10B01-135_A1_BCSA4_TumA1/outs/",
                        slice = "BCSA4TumA1")

# Calculate percentage of ribosomal genes
ribo <- grep("^RP[SL]", rownames(bcsa), value = T)
bcsa[["percent_ribo"]] <- PercentageFeatureSet(bcsa, features = ribo)

# Calculate percentage of mitochondrial genes
bcsa[["percent_mito"]] <- PercentageFeatureSet(object = bcsa, pattern = "^MT-")

# Calculate percentage of hemoglobin genes
hb <- grep("^HB", rownames(bcsa), value = T)
hb <- hb[-which(hb %in% c("HBEGF", "HBS1L", "HBP1"))]
bcsa[["percent_hb"]] <- PercentageFeatureSet(bcsa, features = hb)

# Remove mitochondria genes
bcsa[["Spatial"]]$counts <- bcsa[["Spatial"]]$counts[-grep("^MT-", rownames(bcsa[["Spatial"]]$counts)),]

# Save the seurat object
save(bcsa, file = "~/Desktop/CIIR/data/PAM50Bulk/results/BCSA4TumA1.RData")

# add image slot to the seurat object
imgpath <- "/Users/tili/Desktop/CIIR/data/visium_spatial_out/V10F24-114_A1/outs/spatial/"
img <- Read10X_Image(image.dir = imgpath)

img <- img[Cells(x = bcsa)]
DefaultAssay(img) <- "Spatial"
bcsa@images$BCSA2TumD3 <- img
SpatialFeaturePlot(bcsa, features = "nCount_Spatial")

```

# Prepare files for Stereoscope deconvolution - subset scRNAseq data to 1000 cells/cluster

```{r Stereoscope deconvolution}

sc.ref <- Read10X("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/", gene.column = 1)
# import meta data
meta <- read.csv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/metadata.csv", row.names = 1)
sc.ref <- CreateSeuratObject(sc.ref, meta.data = meta)

# Subset HER2+ data
Idents(sc.ref) <- sc.ref@meta.data$subtype
sc.her2 <- subset(sc.ref, idents = "HER2+")

# Subset TNBC data
sc.tnbc <- subset(sc.ref, idents = "TNBC")

# Required data output: Spatial data: tsv file with spots as rows and genes as columns
#                       Single cell count data: tsv file with cells as rows and genes as variables, unique row names
#                       Single cell annotation: tsv file with same rownames (cellid) with one column listing annotation (colname is bio_celltype), 
#                                               or multiple columns where the conlumn containing the lables should be named as "bio_celltype**"
# Make sure ST and sc data use the same gene identifiers
# Do NOT normalize data, model relies on the raw count data where counts should be integer

# Prepare sc ref data, use minor annotation - HER2+
her2.meta <- sc.her2@meta.data %>% dplyr::select(celltype_minor) %>% 'colnames<-' ("bio_celltype")
her2.meta$bio_celltype <- gsub(" ", "_", her2.meta$bio_celltype)

# Relocate cell column - HER2+
her2.meta$cell <- rownames(her2.meta)
her2.meta <- her2.meta %>% relocate(!!"cell")
her2.mtx <- sc.her2[["RNA"]]$counts

# Randomly sample 1000 cells in each group from each cluster. 
# If cluster has cells < 1000, get the cell id
cellid <- ""
gro <- unique(her2.meta$bio_celltype)

for (i in gro) {
  temp1 <- her2.meta %>% filter(bio_celltype == i)
  if (nrow(temp1) > 1000) {
    inTest <- sample(1:nrow(temp1), 1000, replace = F)
    id <- rownames(temp1)[inTest]
  } else if (nrow(temp1) < 1000) {
    id <- rownames(temp1)
  }
  cellid <- c(cellid, id)
}

# Subset the matrix
her2.mtx <- her2.mtx[, which(colnames(her2.mtx) %in% cellid)]

# Find variable features
#sc.tnbc <- NormalizeData(sc.tnbc, normalization.method = "LogNormalize", scale.factor = 10000)
#sc.tnbc <- FindVariableFeatures(sc.tnbc, selection.method = "vst", nfeatures = 3000)

#features <- VariableFeatures(sc.tnbc) 

# Add cell type markers, top50 DEGs
#markers <- openxlsx::read.xlsx("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/DEGs_minor_celltype.xlsx")
#genes <- markers %>% group_by(cluster) %>% top_n(n = 50, wt = avg_logFC)

#features <- c(features, genes$gene) %>% unique() %>% as.data.frame()
#write_delim(features, file = "~/Desktop/CIIR/data/ST_data/stereoscope/tnbc_features.txt", col_names = F)

#############################

# Prepare sc ref data, use minor annotation - TNBC
tnbc.meta <- sc.tnbc@meta.data %>% dplyr::select(celltype_minor) %>% 'colnames<-' ("bio_celltype")
tnbc.meta$bio_celltype <- gsub(" ", "_", tnbc.meta$bio_celltype)

# Relocate cell column - TNBC
tnbc.meta$cell <- rownames(tnbc.meta)
tnbc.meta <- tnbc.meta %>% relocate(!!"cell")
tnbc.mtx <- sc.tnbc[["RNA"]]$counts

# Randomly sample 1000 cells in each group from each cluster. 
# If cluster has cells < 1000, get the cell id
cellid <- ""
gro <- unique(tnbc.meta$bio_celltype)

for (i in gro) {
  temp1 <- tnbc.meta %>% filter(bio_celltype == i)
  if (nrow(temp1) > 1000) {
    inTest <- sample(1:nrow(temp1), 1000, replace = F)
    id <- rownames(temp1)[inTest]
  } else if (nrow(temp1) < 1000) {
    id <- rownames(temp1)
  }
  cellid <- c(cellid, id)
}

# Subset the matrix
tnbc.mtx <- tnbc.mtx[, which(colnames(tnbc.mtx) %in% cellid)]

##############################

# Prepare Spatial data
info <- openxlsx::read.xlsx("~/Desktop/CIIR/data/ST-AR_Visium_samples_with file names.xlsx") %>% na.omit()
info$id <- paste0(info$Patientid, info$Tumor.area)
info <- info[-which(info$id == "BCSA2TumA1"), ]
info <- info[-which(info$id == "BCSA2TumA2"), ]

for (i in 1:nrow(info)) {
  # Load seurat ST data
  # The Processed data have <25% mitochondrial reads, <20% hb-reads, at least 500 detected genes, remove mitochondria genes
  patientid <- info$id[i]
  load(paste0("/Users/tili/Desktop/CIIR/data/PAM50Bulk/results/", patientid, ".RData"))
  
  # Filter the low quality spots and find highly variable genes
  bcsa <- subset(bcsa, subset = nFeature_Spatial > 500 & percent_mito < 25 & percent_hb < 0.2)
  #features <- bcsa@assays$SCT@var.features
  #write.table(features, file = paste0("/Users/tili/Desktop/CIIR/data/ST_data/", patientid, ".features.txt"),
  #            col.names = F, row.names = F)
    
  temp2 <- bcsa[["Spatial"]]$counts
  
  # Get cell coordinates from ST data
  df <- get(patientid, bcsa@images)
  temp1.coor <- df@coordinates
  temp1.coor$coor <- paste0(temp1.coor$imagerow, "x", temp1.coor$imagecol)
  temp1.coor$cellid <- rownames(temp1.coor)
  temp1.coor <- temp1.coor %>% dplyr::select(cellid, coor)
  
  # Identify the common genes between ST and sc data
  if (info$type[i] == "TNBC") {
    temp3 <- tnbc.mtx
    common <- intersect(rownames(temp2), rownames(temp3)) 
    temp3 <- temp3[which(rownames(temp3) %in% common), ]
    temp3 <- t(temp3) %>% as.data.frame()
    
    # Add rownames to the column and move the column to the first one
    temp3$cell <- rownames(temp3)
    temp3 <- temp3 %>% relocate(!!"cell")
    temp3_meta <- tnbc.meta[which(tnbc.meta$cell %in% rownames(temp3)),]
    
    # Files needed to be saved properly as tsv file, file generated using write.table function will give error when running the code
    # Save sc ref data
    write_tsv(temp3_meta, file = paste0("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/tnbc/latest/", patientid, "_tnbc_meta.tsv"))
    write_tsv(temp3, file = paste0("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/tnbc/latest/", patientid, "_tnbc_mtx.tsv"))
    
  } else if (info$type[i] == "HER2+") {
    temp3 <- her2.mtx
    common <- intersect(rownames(temp2), rownames(her2.mtx))
    temp3 <- temp3[which(rownames(temp3) %in% common),]
    temp3 <- t(temp3) %>% as.data.frame()
    
    # Add rownames to the column and move the column to the first one
    temp3$cell <- rownames(temp3)
    temp3 <- temp3 %>% relocate(!!"cell")
    temp3_meta <- her2.meta[which(her2.meta$cell %in% rownames(temp3)),]
    
    # Files needed to be saved properly as tsv file, file generated using write.table function will give error when running the code
    write_tsv(temp3_meta, file = paste0("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/her2/latest/", patientid, "_her2_meta.tsv"))
    write_tsv(temp3, file = paste0("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/her2/latest/", patientid, "_her2_mtx.tsv"))
  }
  
  # Subset the ST data
  temp2 <- temp2[which(rownames(temp2) %in% common),] %>% t() %>% as.data.frame()
  
  # Merge ST count matrix with coordinate dataframe, remove cell id column and set coordinate as row and genes as column
  temp2$cell <- rownames(temp2)
  temp2 <- merge(temp2, temp1.coor, by.x = "cell", by.y = "cellid")
  temp2 <- temp2 %>% relocate(!!"coor")
  temp2 <- temp2[,-which(colnames(temp2) == "cell")]
  
  # Write tsv file
  write_tsv(temp2, file = paste0("/Users/tili/Desktop/CIIR/data/ST_data/stereoscope/latest/", patientid, ".mtx.tsv"))
}

```

# Use scRNAseq ref data from Wu et al 2021 to deconvolute ST data (Without subset) - Stereoscope

```{r Stereoscope deconvolution subset}

sc.ref <- Read10X("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/", gene.column = 1)
# import meta data
meta <- read.csv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/metadata.csv", row.names = 1)
sc.ref <- CreateSeuratObject(sc.ref, meta.data = meta)

# Subset HER2+ data
Idents(sc.ref) <- sc.ref@meta.data$subtype
sc.her2 <- subset(sc.ref, idents = "HER2+")

# Subset TNBC data
sc.tnbc <- subset(sc.ref, idents = "TNBC")

# Required data output: Spatial data: tsv file with spots as rows and genes as columns
#                       Single cell count data: tsv file with cells as rows and genes as variables, unique row names
#                       Single cell annotation: tsv file with same rownames (cellid) with one column listing annotation (colname is bio_celltype), 
#                                               or multiple columns where the conlumn containing the lables should be named as "bio_celltype**"
# Make sure ST and sc data use the same gene identifiers
# Do NOT normalize data, model relies on the raw count data where counts should be integer

# Prepare sc ref data, use minor annotation - HER2+
her2.meta <- sc.her2@meta.data %>% dplyr::select(celltype_minor) %>% 'colnames<-' ("bio_celltype")
her2.meta$bio_celltype <- gsub(" ", "_", her2.meta$bio_celltype)

# Relocate cell column - HER2+
her2.meta$cell <- rownames(her2.meta)
her2.meta <- her2.meta %>% relocate(!!"cell")
her2.mtx <- sc.her2[["RNA"]]$counts

#############################

# Prepare sc ref data, use minor annotation - TNBC
tnbc.meta <- sc.tnbc@meta.data %>% dplyr::select(celltype_minor) %>% 'colnames<-' ("bio_celltype")
tnbc.meta$bio_celltype <- gsub(" ", "_", tnbc.meta$bio_celltype)

# Relocate cell column - TNBC
tnbc.meta$cell <- rownames(tnbc.meta)
tnbc.meta <- tnbc.meta %>% relocate(!!"cell")
tnbc.mtx <- sc.tnbc[["RNA"]]$counts

##############################

# Prepare Spatial data
info <- openxlsx::read.xlsx("~/Desktop/CIIR/data/ST-AR_Visium_samples_with file names.xlsx") %>% na.omit()
info$id <- paste0(info$Patientid, info$Tumor.area)
info <- info[-which(info$id == "BCSA2TumA1"), ]
info <- info[-which(info$id == "BCSA2TumA2"), ]

for (i in 1:nrow(info)) {
  # Load seurat ST data
  patientid <- info$id[i]
  load(paste0("/Users/tili/Desktop/CIIR/data/PAM50Bulk/results/", patientid, ".RData"))
  temp2 <- bcsa[["Spatial"]]$counts
  
  # Get cell coordinates from ST data
  df <- get(patientid, bcsa@images)
  temp1.coor <- df@coordinates
  temp1.coor$coor <- paste0(temp1.coor$imagerow, "x", temp1.coor$imagecol)
  temp1.coor$cellid <- rownames(temp1.coor)
  temp1.coor <- temp1.coor %>% dplyr::select(cellid, coor)
  
  # Identify the common genes between ST and sc data
  if (info$type[i] == "TNBC") {
    common <- intersect(rownames(temp2), rownames(tnbc.mtx)) 
  } else if (info$type[i] == "HER2+") {
    common <- intersect(rownames(temp2), rownames(her2.mtx))
  }
  
  # Subset the ST data
  temp2 <- temp2[which(rownames(temp2) %in% common),] %>% t() %>% as.data.frame()
  
  # Merge ST count matrix with coordinate dataframe, remove cell id column and set coordinate as row and genes as column
  temp2$cell <- rownames(temp2)
  temp2 <- merge(temp2, temp1.coor, by.x = "cell", by.y = "cellid")
  temp2 <- temp2 %>% relocate(!!"coor")
  temp2 <- temp2[,-which(colnames(temp2) == "cell")]
  
  # Write csv file
  write_tsv(temp2, file = paste0("/Users/tili/Desktop/CIIR/data/ST_data/", patientid, ".mtx.tsv"))
}

#patientid <- info$id[1]
#load(paste0("/Users/tili/Desktop/CIIR/data/PAM50Bulk/results/", patientid, ".RData"))
#bcsa1.a1.mtx <- bcsa[["Spatial"]]$counts

# Get cell coordinates from ST data
#df <- get(patientid, bcsa@images)
#temp1.coor <- df@coordinates
#temp1.coor$coor <- paste0(temp1.coor$imagerow, "x", temp1.coor$imagecol)
#temp1.coor$cellid <- rownames(temp1.coor)
#temp1.coor <- temp1.coor %>% dplyr::select(cellid, coor)

# Keep the common genes between ref dataset and ST data
common <- intersect(rownames(bcsa3.c1.mtx), rownames(her2.mtx))
#bcsa3.c2.mtx <- bcsa3.c2.mtx[which(rownames(bcsa3.c2.mtx) %in% common),] %>% t() %>% as.data.frame()
her2.mtx <- her2.mtx[which(rownames(her2.mtx) %in% common),]
her2.mtx <- t(her2.mtx) %>% as.data.frame()

# Add rownames to the column and move the column to the first one
her2.mtx$cell <- rownames(her2.mtx)
her2.mtx <- her2.mtx %>% relocate(!!"cell")

############################## TNBC

# Keep the common genes between ref dataset and ST data
common <- intersect(rownames(bcsa1.a1.mtx), rownames(tnbc.mtx))
#bcsa1.a1.mtx <- bcsa1.a1.mtx[which(rownames(bcsa1.a1.mtx) %in% common),] %>% t() %>% as.data.frame()
tnbc.mtx <- tnbc.mtx[which(rownames(tnbc.mtx) %in% common),]
tnbc.mtx <- t(tnbc.mtx) %>% as.data.frame()

# Add rownames to the column and move the column to the first one
tnbc.mtx$cell <- rownames(tnbc.mtx)
tnbc.mtx <- tnbc.mtx %>% relocate(!!"cell")

# Merge ST count matrix with coordinate dataframe, remove cell id column and set coordinate as row and genes as column
#bcsa3.c2.mtx$cell <- rownames(bcsa3.c2.mtx)
#bcsa3.c2.mtx <- merge(bcsa3.c2.mtx, temp1.coor, by.x = "cell", by.y = "cellid")
#bcsa3.c2.mtx <- bcsa3.c2.mtx %>% relocate(!!"coor")
#bcsa3.c2.mtx <- bcsa3.c2.mtx[,-which(colnames(bcsa3.c2.mtx) == "cell")]

# Files needed to be saved properly as tsv file, file generated using write.table function will give error when running the code
write_tsv(her2.meta, file = "/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/her2/her2_meta.tsv")
write_tsv(her2.mtx, file = "/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/her2/her2_mtx.tsv")

write_tsv(tnbc.meta, file = "/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/tnbc/tnbc_meta.tsv")
write_tsv(tnbc.mtx, file = "/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/tnbc/tnbc_mtx.tsv")

#write_tsv(bcsa3.c2.mtx, file = "/Users/tili/Desktop/CIIR/data/ST_data/bcsa3.c2.mtx.tsv")

```


# Prepare files for CytoSpace deconvolution - subset scRNAseq data to 1000 cells/cluster 

```{r cytospace deconvolution}

# By default, CytoSPACE requires 4 files as input. All files should be provided in a tab-delimited tabular input format (.txt) with no double quotations.

# scRNAseq matrix: genes as rows and cells as columns; first row must contain singel cell IDs and first column contains gene names; first column 
# must have a header; data should be non-normalized counts; duplicate gene names will be dropped

# Cell type annotation file: annotation corresponding to cell ID; label strings should not have special characters; table should contain two
# columns, column 1 is cell id corresponding to the columns of matrix and 2 contains annotations; columns much have a header

# ST matrix file: genes as rows and ST spots as columns; first row contains the ST spot ID and first column gene names; first column must have a 
# header; data should be non-normalized counts; duplicate will be dropped

# ST coordinate file: a table contains 3 columns, first column contains the ST spot ID corresponding to the columns of ST matrix, 2 and 3 contain
# the row and column indices of each ST spot; columns must have a header

# Input can be generated from spaceranger output or from a seurat object using R code in the prepare_input_files folder
source("/Users/tili/Desktop/CIIR/codes/cytospace/cytospace/Prepare_input_files/generate_cytospace_from_seurat_object.R")

# Use the cells selected and used in stereoscope
tnbc.cellid <- read_tsv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/tnbc/BCSA1TumA1_tnbc_meta.tsv")
tnbc.cellid <- tnbc.cellid$cell

# Get the common genes between sc and st data
common <- read_tsv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/tnbc/BCSA1TumA1_tnbc_mtx.tsv")
tnbc_common <- colnames(common)[-1]

her2.cellid <- read_tsv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/her2/BCSA2TumB1_her2_meta.tsv")
her2.cellid <- her2.cellid$cell

# Get the common genes between sc and st data
common <- read_tsv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/her2/BCSA2TumB1_her2_mtx.tsv")
her2_common <- colnames(common)[-1]

# Subset the seurat object to cells used in stereoscope
tnbc.mtx <- sc.tnbc
her2.mtx <- sc.her2

tnbc.mtx <- tnbc.mtx[which(rownames(tnbc.mtx) %in% tnbc_common), which(colnames(tnbc.mtx) %in% tnbc.cellid)]
her2.mtx <- her2.mtx[which(rownames(her2.mtx) %in% her2_common), which(colnames(her2.mtx) %in% her2.cellid)]

# Generate CytoSPACE inputs from Seurat objects, Idents(seurat_object) must be set to include cell types
Idents(tnbc.mtx) <- tnbc.mtx@meta.data$celltype_minor
Idents(her2.mtx) <- her2.mtx@meta.data$celltype_minor

generate_cytospace_from_scRNA_seurat_object(tnbc.mtx, dir_out = "/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/cytospace/tnbc/latest/",
                                            fout_prefix = "tnbc_", rna_assay = "RNA")

generate_cytospace_from_scRNA_seurat_object(her2.mtx, dir_out = "/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/cytospace/her2/latest/",
                                            fout_prefix = "her2_", rna_assay = "RNA")

##############################

# Prepare Spatial data
info <- openxlsx::read.xlsx("~/Desktop/CIIR/data/ST-AR_Visium_samples_with file names.xlsx") %>% na.omit()
info$id <- paste0(info$Patientid, info$Tumor.area)
info <- info[-which(info$id == "BCSA2TumA1"), ]
info <- info[-which(info$id == "BCSA2TumA2"), ]

for (i in 1:nrow(info)) {
  # Load seurat ST data
  # The Processed data have <25% mitochondrial reads, <20% hb-reads, at least 500 detected genes, remove mitochondria genes
  patientid <- info$id[i]
  load(paste0("/Users/tili/Desktop/CIIR/data/PAM50Bulk/results/", patientid, ".RData"))
  
  # Filter the low quality spots and find highly variable genes
  bcsa <- subset(bcsa, subset = nFeature_Spatial > 500 & percent_mito < 25 & percent_hb < 0.2)
  
  # keep the common genes between sc and st
  if (info$type[i] == "TNBC") {
    bcsa[["Spatial"]]$counts <- bcsa[["Spatial"]]$counts[which(rownames(bcsa[["Spatial"]]$counts) %in% tnbc_common),]
  } else {
    bcsa[["Spatial"]]$counts <- bcsa[["Spatial"]]$counts[which(rownames(bcsa[["Spatial"]]$counts) %in% her2_common),]
  }
  
  
  # Generate Spatial data, slice name should be given (patientid in this case) to be able to extract the coordinates
  generate_cytospace_from_ST_seurat_object(bcsa, dir_out='/Users/tili/Desktop/CIIR/data/ST_data/cytospace/latest/', 
                                           fout_prefix= paste0(patientid, "_"), write_sparse=FALSE, slice = patientid)
}


```

# Use CARD for ST deconvolution

```{r CARD}

# CARD requires two inputs: ST count data, along with coordinates; scRNAseq count data and meta information indicating cell types and sample info

sc.ref <- Read10X("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/", gene.column = 1)
# import meta data
meta <- read.csv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/metadata.csv", row.names = 1)
sc.ref <- CreateSeuratObject(sc.ref, meta.data = meta)

# Subset HER2+ data
Idents(sc.ref) <- sc.ref@meta.data$subtype
sc.her2 <- subset(sc.ref, idents = "HER2+")

# Subset TNBC data
sc.tnbc <- subset(sc.ref, idents = "TNBC")

# Use the cells selected and used in stereoscope
tnbc.cellid <- read_tsv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/tnbc/BCSA1TumA1_tnbc_meta.tsv")
tnbc.cellid <- tnbc.cellid$cell

her2.cellid <- read_tsv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/her2/BCSA2TumB1_her2_meta.tsv")
her2.cellid <- her2.cellid$cell

# Subset the seurat object to cells used in stereoscope
tnbc.mtx <- sc.tnbc
her2.mtx <- sc.her2

tnbc.mtx <- tnbc.mtx[,which(colnames(tnbc.mtx) %in% tnbc.cellid)]
her2.mtx <- her2.mtx[,which(colnames(her2.mtx) %in% her2.cellid)]

################# Prepare ST data

for (i in 1:nrow(info)) {
  # ST count data must be matrix or sparseMatrix, row represents a gene and column represent a spatial location (can be spot coor or cell id)
  patientid <- info$id[i]
  load(paste0("/Users/tili/Desktop/CIIR/data/PAM50Bulk/results/", patientid, ".RData"))
  
  # Filter the low quality spots and find highly variable genes
  bcsa <- subset(bcsa, subset = nFeature_Spatial > 500 & percent_mito < 25 & percent_hb < 0.2)
  st_count <- bcsa[["Spatial"]]$counts

  # ST metadata: must be data frame while each row represents a spatial location, first column is x coor and second is y coor. The rownames of the
  # spatial location data frame should match exactly with the column names of the count matrix, the colname should be x and y
  df <- get(patientid, bcsa@images)
  temp1.coor <- df@coordinates
  temp1.coor$coor <- paste0(temp1.coor$imagerow, "x", temp1.coor$imagecol)
  temp1.coor$cellid <- rownames(temp1.coor)
  st_coor <- temp1.coor %>% dplyr::select(row, col)
  colnames(st_coor) <- c("x", "y")

  ################### Prepare scRNAseq data

  if (info$type[i] == "TNBC") {
    # scRNAseq count data must be matrix or sparseMatrix, each row represents a gene and column a cell
    sc_count <- tnbc.mtx[["RNA"]]$counts
    
    # Keep only the common genes between ST and sc data
    common <- intersect(rownames(sc_count), rownames(st_count))
    sc_count <- sc_count[which(rownames(sc_count) %in% common),]
    st_count <- st_count[which(rownames(st_count) %in% common),]

    # scRNAseq meta data must be data frame while each row represents a cells, 
    # rownames of meta data should match exactly with column names of sc_count
    # the sc_meta data must contain a column indicating the cell type, sample information, 
    # the input of scRNAseq in each cell type should be at least 2
    sc_meta <- tnbc.mtx@meta.data
    sc_meta$cellID <- rownames(sc_meta)
    sc_meta <- sc_meta %>% dplyr::select(cellID, orig.ident, celltype_minor)
    
  } else if (info$type[i] == "HER2+") {
    # scRNAseq count data must be matrix or sparseMatrix, each row represents a gene and column a cell
    sc_count <- her2.mtx[["RNA"]]$counts
    
    # Keep only the common genes between ST and sc data
    common <- intersect(rownames(sc_count), rownames(st_count))
    sc_count <- sc_count[which(rownames(sc_count) %in% common),]
    st_count <- st_count[which(rownames(st_count) %in% common),]
    
    sc_meta <- her2.mtx@meta.data
    sc_meta$cellID <- rownames(sc_meta)
    sc_meta <- sc_meta %>% dplyr::select(cellID, orig.ident, celltype_minor)
  }

#################### Running CARD deconvolution

  # Create an CARD object
  # The spatial data are stored in CARD_obj@spatial_countMat and CARD_obj@spatial_location while 
  # the scRNA-seq data is stored in CARD_obj@sc_eset in the format of SingleCellExperiment.
  CARD_obj = createCARDObject(
	  sc_count = sc_count,
	  sc_meta = sc_meta,
	  spatial_count = st_count,
	  spatial_location = st_coor,
	  ct.varname = "celltype_minor",
	  ct.select = unique(sc_meta$cellType),
	  sample.varname = "orig.ident",
	  minCountGene = 100,
	  minCountSpot = 5) 

  # Deconvolution using CARD
  CARD_obj = CARD_deconvolution(CARD_object = CARD_obj)

  # Extract the results
  temp1.card <- CARD_obj@Proportion_CARD %>% as.data.frame()
  write.csv(temp1.card, file = paste0("/Users/tili/Desktop/CIIR/results/CARD/", patientid, "_CARD_subset_1000_minor_common_genes.csv"))
}

```

# Tangram and cell2location deconvolution

```{r Tangram and cell2location}

sc.ref <- Read10X("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/", gene.column = 1)
# import meta data
meta <- read.csv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/metadata.csv", row.names = 1)
sc.ref <- CreateSeuratObject(sc.ref, meta.data = meta)

# Subset HER2+ data
Idents(sc.ref) <- sc.ref@meta.data$subtype
sc.her2 <- subset(sc.ref, idents = "HER2+")

# Subset TNBC data
sc.tnbc <- subset(sc.ref, idents = "TNBC")

# Use the cells selected and used in stereoscope
tnbc.cellid <- read_tsv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/tnbc/BCSA1TumA1_tnbc_meta.tsv")
tnbc.cellid <- tnbc.cellid$cell

her2.cellid <- read_tsv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/her2/BCSA2TumB1_her2_meta.tsv")
her2.cellid <- her2.cellid$cell

# Subset the seurat object to cells used in stereoscope
tnbc.mtx <- sc.tnbc
her2.mtx <- sc.her2

tnbc.mtx <- tnbc.mtx[,which(colnames(tnbc.mtx) %in% tnbc.cellid)]
her2.mtx <- her2.mtx[,which(colnames(her2.mtx) %in% her2.cellid)]

# Set the identity to cell type annotation
Idents(tnbc.mtx) <- tnbc.mtx@meta.data$celltype_minor
Idents(her2.mtx) <- her2.mtx@meta.data$celltype_minor

# Switch RNA Assay5 class into Assay class before convert to h5Seurat
tnbc.mtx[["RNA"]] <- as(object = tnbc.mtx[["RNA"]], Class = "Assay")
her2.mtx[["RNA"]] <- as(object = her2.mtx[["RNA"]], Class = "Assay")

# Get the common genes 
common <- read_tsv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/tnbc/BCSA1TumA1_tnbc_mtx.tsv")
common <- read_tsv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/her2/BCSA2TumB1_her2_mtx.tsv")
common <- colnames(common)[-1]

tnbc.mtx <- tnbc.mtx[which(rownames(tnbc.mtx) %in% common), ]
her2.mtx <- her2.mtx[which(rownames(her2.mtx) %in% common), ]

# Save the reference data
SaveH5Seurat(tnbc.mtx, filename = "/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/Tangram/latest/tnbc.h5Seurat", overwrite = T, verbose = T)
Convert("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/Tangram/latest/tnbc.h5Seurat", dest = "h5ad", overwrite = T, verbose = T)

SaveH5Seurat(her2.mtx, filename = "/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/Tangram/latest/her2.h5Seurat", overwrite = T, verbose = T)
Convert("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/Tangram/latest/her2.h5Seurat", dest = "h5ad", overwrite = T, verbose = T)

##############################

# Prepare Spatial data
info <- openxlsx::read.xlsx("~/Desktop/CIIR/data/ST-AR_Visium_samples_with file names.xlsx") %>% na.omit()
info$id <- paste0(info$Patientid, info$Tumor.area)
info <- info[-which(info$id == "BCSA2TumA1"), ]
info <- info[-which(info$id == "BCSA2TumA2"), ]

# Convert ST Seurat to h5ad format for tangram analysis
for (i in 1:nrow(info)) {
  patientid <- info$id[i]
  # Use all the genes instead of the top 3000 variable genes from SCTransformation
  load(paste0("/Users/tili/Desktop/CIIR/data/PAM50Bulk/results/", patientid, ".RData"))
  
  # Filter the low quality spots and find highly variable genes
  bcsa <- subset(bcsa, subset = nFeature_Spatial > 500 & percent_mito < 25 & percent_hb < 0.2)
  
  if (info$type[i] == "TNBC") {
    common <- intersect(rownames(bcsa), rownames(tnbc.mtx)) 
    bcsa <- bcsa[which(rownames(bcsa) %in% common), ]
  } else if (info$type[i] == "HER2+") {
    common <- intersect(rownames(bcsa), rownames(her2.mtx)) 
    bcsa <- bcsa[which(rownames(bcsa) %in% common), ]
  }
  
  # Switch RNA Assay5 class into Assay class before convert to h5Seurat
  bcsa[["Spatial"]] <- as(object = bcsa[["Spatial"]], Class = "Assay")
  
  # Tangram data preparation
  SaveH5Seurat(bcsa, filename = paste0("/Users/tili/Desktop/CIIR/data/ST_data/Tangram/latest/", patientid, ".h5Seurat"))
  Convert(paste0("/Users/tili/Desktop/CIIR/data/ST_data/Tangram/latest/", patientid, ".h5Seurat"), dest = "h5ad")
}
  
```

## Get marker genes for each cell type for Tangram

```{r get marker genes}

markers <- openxlsx::read.xlsx("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/DEGs_minor_celltype.xlsx")

genes <- markers %>% group_by(cluster) %>% top_n(n = 50, wt = avg_logFC)
write.csv(genes, file = "/Users/tili/Desktop/CIIR/data/ST_data/Tangram/features.csv")

```

# Validation

## Read spaceranger output and create a seurat object

```{r create seurat object}

# Load 10x spaceranger output
bcsa <- Load10X_Spatial(data.dir = "~/Desktop/CIIR/data/visium_spatial_out/V10B01-135_A1/outs/",
                        slice = "BCSA4TumA1")

# Calculate percentage of ribosomal genes
ribo <- grep("^RP[SL]", rownames(bcsa), value = T)
bcsa[["percent_ribo"]] <- PercentageFeatureSet(bcsa, features = ribo)

# Calculate percentage of mitochondrial genes
bcsa[["percent_mito"]] <- PercentageFeatureSet(object = bcsa, pattern = "^MT-")

# Calculate percentage of hemoglobin genes
hb <- grep("^HB", rownames(bcsa), value = T)
hb <- hb[-which(hb %in% c("HBEGF", "HBS1L", "HBP1"))]
bcsa[["percent_hb"]] <- PercentageFeatureSet(bcsa, features = hb)

# Remove mitochondria genes
bcsa[["Spatial"]]$counts <- bcsa[["Spatial"]]$counts[-grep("^MT-", rownames(bcsa[["Spatial"]]$counts)),]

# Save the seurat object
save(bcsa, file = "~/Desktop/CIIR/data/PAM50Bulk/results/BCSA4TumA1.RData")

############## HER2 paper datasets

# Load 10x spaceranger output
for (i in c("A1", "B1", "C1", "D1", "E1", "F1", "G1", "H1")) {
  bcsa <- read.table(paste0("~/Desktop/CIIR/data/Validation_10x_dataset/validation/", i, ".tsv"), 
                     sep = "\t", header = T, row.names = 1) %>% t()
  bcsa <- CreateSeuratObject(counts = bcsa, assay = "Spatial")

  # Calculate percentage of ribosomal genes
  ribo <- grep("^RP[SL]", rownames(bcsa), value = T)
  bcsa[["percent_ribo"]] <- PercentageFeatureSet(bcsa, features = ribo)

  # Calculate percentage of mitochondrial genes
  bcsa[["percent_mito"]] <- PercentageFeatureSet(object = bcsa, pattern = "^MT-")

  # Calculate percentage of hemoglobin genes
  hb <- grep("^HB", rownames(bcsa), value = T)
  hb <- hb[-which(hb %in% c("HBEGF", "HBS1L", "HBP1"))]
  bcsa[["percent_hb"]] <- PercentageFeatureSet(bcsa, features = hb)

  # Remove mitochondria genes
  if (length(grep("^MT-", rownames(bcsa[["Spatial"]]$counts))) > 0) {
    bcsa[["Spatial"]]$counts <- bcsa[["Spatial"]]$counts[-grep("^MT-", rownames(bcsa[["Spatial"]]$counts)),]
  }

  # Save the seurat object
  save(bcsa, file = paste0("~/Desktop/CIIR/data/PAM50Bulk/results/", i, ".RData"))
}


```

# Cell2location deconvolution

```{r Cell2location}

sc.ref <- Read10X("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/", gene.column = 1)
# import meta data
meta <- read.csv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/metadata.csv", row.names = 1)
sc.ref <- CreateSeuratObject(sc.ref, meta.data = meta)

# Subset HER2+ data
Idents(sc.ref) <- sc.ref@meta.data$subtype
sc.her2 <- subset(sc.ref, idents = "HER2+")

# Use the cells selected and used in stereoscope
her2.cellid <- read_tsv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/her2/BCSA2TumB1_her2_meta.tsv")
her2.cellid <- her2.cellid$cell

# Subset the seurat object to cells used in stereoscope
her2.mtx <- sc.her2

her2.mtx <- her2.mtx[,which(colnames(her2.mtx) %in% her2.cellid)]

# Set the identity to cell type annotation
Idents(her2.mtx) <- her2.mtx@meta.data$celltype_minor

# Switch RNA Assay5 class into Assay class before convert to h5Seurat
her2.mtx[["RNA"]] <- as(object = her2.mtx[["RNA"]], Class = "Assay")

################### Get the common genes for BCSA4TumA1

load("~/Desktop/CIIR/data/PAM50Bulk/results/BCSA4TumA1.RData")
common <- intersect(rownames(bcsa), rownames(her2.mtx))

her2.mtx <- her2.mtx[which(rownames(her2.mtx) %in% common), ]

SaveH5Seurat(her2.mtx, filename = "/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/cell2location_validation/her2_BCSA4TumA1.h5Seurat", overwrite = T, verbose = T)
Convert("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/cell2location_validation/her2_BCSA4TumA1.h5Seurat", 
        dest = "h5ad", overwrite = T, verbose = T)

################## Get the common genes for HER2 paper datasets

for (i in c("A1", "B1", "C1", "D1", "E1", "F1", "H1", "G1")) {
 # Subset the seurat object to cells used in stereoscope
  her2.mtx <- sc.her2
  her2.mtx <- her2.mtx[,which(colnames(her2.mtx) %in% her2.cellid)]

  # Set the identity to cell type annotation
  Idents(her2.mtx) <- her2.mtx@meta.data$celltype_minor

  # Switch RNA Assay5 class into Assay class before convert to h5Seurat
  her2.mtx[["RNA"]] <- as(object = her2.mtx[["RNA"]], Class = "Assay")

  load(paste0("~/Desktop/CIIR/data/PAM50Bulk/results/", i, ".RData"))
  common <- intersect(rownames(bcsa), rownames(her2.mtx))

  her2.mtx <- her2.mtx[which(rownames(her2.mtx) %in% common), ]

  # Save the reference data
  SaveH5Seurat(her2.mtx, filename = paste0("~/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/cell2location_validation/her2_",
                i, ".h5Seurat"), 
               overwrite = T, verbose = T)
  Convert(paste0("~/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/cell2location_validation/her2_", i, ".h5Seurat"), 
        dest = "h5ad", overwrite = T, verbose = T) 
}


############################## Prepare ST data

# Subset the seurat object to cells used in stereoscope
her2.mtx <- sc.her2

her2.mtx <- her2.mtx[,which(colnames(her2.mtx) %in% her2.cellid)]

# Set the identity to cell type annotation
Idents(her2.mtx) <- her2.mtx@meta.data$celltype_minor

# Switch RNA Assay5 class into Assay class before convert to h5Seurat
her2.mtx[["RNA"]] <- as(object = her2.mtx[["RNA"]], Class = "Assay")

# Convert ST Seurat to h5ad format for cell2location analysis
for (i in c("A1", "B1", "C1", "D1", "E1", "F1", "H1", "G1", "BCSA4TumA1")) {
  # Use all the genes instead of the top 3000 variable genes from SCTransformation
  load(paste0("/Users/tili/Desktop/CIIR/data/PAM50Bulk/results/", i, ".RData"))
  
  # Filter the low quality spots and find highly variable genes
  bcsa <- subset(bcsa, subset = nFeature_Spatial > 500 & percent_mito < 25 & percent_hb < 0.2)
  
  common <- intersect(rownames(bcsa), rownames(her2.mtx)) 
  bcsa <- bcsa[which(rownames(bcsa) %in% common), ]
  
  # Switch RNA Assay5 class into Assay class before convert to h5Seurat
  bcsa[["Spatial"]] <- as(object = bcsa[["Spatial"]], Class = "Assay")
  
  # Tangram data preparation
  SaveH5Seurat(bcsa, filename = paste0("/Users/tili/Desktop/CIIR/data/ST_data/cell2location_validation/", i, ".h5Seurat"))
  Convert(paste0("/Users/tili/Desktop/CIIR/data/ST_data/cell2location_validation/", i, ".h5Seurat"), dest = "h5ad")
}
  
```

# DWLS and RCTD

```{r DWLS and RCTD}

sc.ref <- Read10X("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/", gene.column = 1)
# import meta data
meta <- read.csv("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/metadata.csv", row.names = 1)
sc.ref <- CreateSeuratObject(sc.ref, meta.data = meta)

# Subset HER2+ data
Idents(sc.ref) <- sc.ref@meta.data$subtype
sc.her2 <- subset(sc.ref, idents = "HER2+")

# Subset TNBC data
sc.tnbc <- subset(sc.ref, idents = "TNBC")

# Required data output: Spatial data: tsv file with spots as rows and genes as columns
#                       Single cell count data: tsv file with cells as rows and genes as variables, unique row names
#                       Single cell annotation: tsv file with same rownames (cellid) with one column listing annotation (colname is bio_celltype), 
#                                               or multiple columns where the conlumn containing the lables should be named as "bio_celltype**"
# Make sure ST and sc data use the same gene identifiers
# Do NOT normalize data, model relies on the raw count data where counts should be integer

# Prepare sc ref data, use minor annotation - HER2+
her2.meta <- sc.her2@meta.data %>% dplyr::select(celltype_minor) %>% 'colnames<-' ("bio_celltype")
her2.meta$bio_celltype <- gsub(" ", "_", her2.meta$bio_celltype)

# Relocate cell column - HER2+
her2.meta$cell <- rownames(her2.meta)
her2.meta <- her2.meta %>% relocate(!!"cell")
her2.mtx <- sc.her2[["RNA"]]$counts

# Subset the matrix
her2.mtx <- her2.mtx[, which(colnames(her2.mtx) %in% her2.cellid)]

#############################

for (i in 1:nrow(info)) {
  # Load seurat ST data
  # The Processed data have <25% mitochondrial reads, <20% hb-reads, at least 500 detected genes, remove mitochondria genes
  patientid <- info$id[i]
  load(paste0("/Users/tili/Desktop/CIIR/data/PAM50Bulk/results/", i, ".RData"))
  # Filter the low quality spots and find highly variable genes
  bcsa <- subset(bcsa, subset = nFeature_Spatial > 500 & percent_mito < 25 & percent_hb < 0.2)
  #features <- bcsa@assays$SCT@var.features
  #write.table(features, file = paste0("/Users/tili/Desktop/CIIR/data/ST_data/", patientid, ".features.txt"),
  #            col.names = F, row.names = F)
    
  temp2 <- bcsa[["Spatial"]]$counts
  
  # Get cell coordinates from ST data
  df <- get(patientid, bcsa@images)
  temp1.coor <- df@coordinates
  temp1.coor$coor <- paste0(temp1.coor$imagerow, "x", temp1.coor$imagecol)
  temp1.coor$cellid <- rownames(temp1.coor)
  temp1.coor <- temp1.coor %>% dplyr::select(cellid, coor)
  
  # Identify the common genes between ST and sc data
  temp3 <- her2.mtx
  common <- intersect(rownames(temp2), rownames(her2.mtx))
  temp3 <- temp3[which(rownames(temp3) %in% common),]
  temp3 <- t(temp3) %>% as.data.frame()
    
  # Add rownames to the column and move the column to the first one
  temp3$cell <- rownames(temp3)
  temp3 <- temp3 %>% relocate(!!"cell")
  temp3_meta <- her2.meta[which(her2.meta$cell %in% rownames(temp3)),]
    
  # Files needed to be saved properly as tsv file, file generated using write.table function will give error when running the code
  write_tsv(temp3_meta, file = paste0("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/her2/", patientid, "_her2_meta.tsv"))
  write_tsv(temp3, file = paste0("/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/stereoscope/her2/", patientid, "_her2_mtx.tsv"))
  
  # Subset the ST data
  temp2 <- temp2[which(rownames(temp2) %in% common),] %>% t() %>% as.data.frame()
  
  # Merge ST count matrix with coordinate dataframe, remove cell id column and set coordinate as row and genes as column
  temp2$cell <- rownames(temp2)
  temp2 <- merge(temp2, temp1.coor, by.x = "cell", by.y = "cellid")
  temp2 <- temp2 %>% relocate(!!"coor")
  temp2 <- temp2[,-which(colnames(temp2) == "cell")]
  
  # Write tsv file
  write_tsv(temp2, file = paste0("/Users/tili/Desktop/CIIR/data/ST_data/stereoscope/", patientid, ".mtx.tsv"))
}

```

# Session info

```{r session info}
sessionInfo()
```
