{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "79733fdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import torch\n",
    "sys.path.append('./')  # uncomment for local import\n",
    "import tangram as tg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d3fa105c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.0.4'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "\n",
    "tg.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c2e02f09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Patientid</th>\n",
       "      <th>Tumor area</th>\n",
       "      <th>Visium ID</th>\n",
       "      <th>Visum h5ad ID</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>width</th>\n",
       "      <th>height</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TD-2532-BCSA1</td>\n",
       "      <td>BCSA1</td>\n",
       "      <td>TumA1</td>\n",
       "      <td>V19T26-012_A1</td>\n",
       "      <td>200116_ST-AR_BC_A_B_V19T26-012_KT_V1-Spot00000...</td>\n",
       "      <td>a1.012</td>\n",
       "      <td>TNBC</td>\n",
       "      <td>29184</td>\n",
       "      <td>29696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TD-2532-BCSA1</td>\n",
       "      <td>BCSA1</td>\n",
       "      <td>TumA2</td>\n",
       "      <td>V19T26-012_B1</td>\n",
       "      <td>200116_ST-AR_BC_A_B_V19T26-012_KT_V2-Spot00000...</td>\n",
       "      <td>b1.012</td>\n",
       "      <td>TNBC</td>\n",
       "      <td>29184</td>\n",
       "      <td>29696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TD-2532-BCSA1</td>\n",
       "      <td>BCSA1</td>\n",
       "      <td>TumB1</td>\n",
       "      <td>V19T26-012_C1</td>\n",
       "      <td>200116_ST-AR_BC_A_B_V19T26-012_KT_V3-Spot00000...</td>\n",
       "      <td>c1.012</td>\n",
       "      <td>TNBC</td>\n",
       "      <td>29184</td>\n",
       "      <td>29696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TD-2532-BCSA1</td>\n",
       "      <td>BCSA1</td>\n",
       "      <td>TumB2</td>\n",
       "      <td>V19T26-012_D1</td>\n",
       "      <td>200116_ST-AR_BC_A_B_V19T26-012_KT_V4-Spot00000...</td>\n",
       "      <td>d1.012</td>\n",
       "      <td>TNBC</td>\n",
       "      <td>29184</td>\n",
       "      <td>29696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumB1</td>\n",
       "      <td>V10F24-112_C1</td>\n",
       "      <td>200616_BC-SA2_V10F24-112_KT.C1-Spot000001.h5ad</td>\n",
       "      <td>c1.112</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>47616</td>\n",
       "      <td>48128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumB2</td>\n",
       "      <td>V10F24-112_D1</td>\n",
       "      <td>200616_BC-SA2_V10F24-112_KT.D1-Spot000001.h5ad</td>\n",
       "      <td>d1.112</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>47616</td>\n",
       "      <td>48128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumC1</td>\n",
       "      <td>V10F24-113_A1</td>\n",
       "      <td>200715_BC_SA2_C_D_V10F24-113.A1-Spot000001.h5ad</td>\n",
       "      <td>a1.113</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumC2</td>\n",
       "      <td>V10F24-113_B1</td>\n",
       "      <td>200715_BC_SA2_C_D_V10F24-113.B1-Spot000001.h5ad</td>\n",
       "      <td>b1.113</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumD1</td>\n",
       "      <td>V10F24-113_C1</td>\n",
       "      <td>200715_BC_SA2_C_D_V10F24-113.C1-Spot000001.h5ad</td>\n",
       "      <td>c1.113</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumD2</td>\n",
       "      <td>V10F24-113_D1</td>\n",
       "      <td>200715_BC_SA2_C_D_V10F24-113.D1-Spot000001.h5ad</td>\n",
       "      <td>d1.113</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumE1</td>\n",
       "      <td>V10F24-114_C1</td>\n",
       "      <td>200715_BC_SA2_D_E_V10F24-114.C1-Spot000001.h5ad</td>\n",
       "      <td>c1.114</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumE2</td>\n",
       "      <td>V10F24-114_D1</td>\n",
       "      <td>200715_BC_SA2_D_E_V10F24-114.D1-Spot000001.h5ad</td>\n",
       "      <td>d1.114</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumA1</td>\n",
       "      <td>V19T26-031_D1</td>\n",
       "      <td>200219_BC_SA3_V19T26-031_CE_V4-Spot000001.h5ad</td>\n",
       "      <td>d1.031</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumA2</td>\n",
       "      <td>V19T26-031_C1</td>\n",
       "      <td>200219_BC_SA3_V19T26-031_CE_V3-Spot000001.h5ad</td>\n",
       "      <td>c1.031</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumB1</td>\n",
       "      <td>V19T26-031_B1</td>\n",
       "      <td>200219_BC_SA3_V19T26-031_CE_V2-Spot000001.h5ad</td>\n",
       "      <td>b1.031</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumB2</td>\n",
       "      <td>V19T26-031_A1</td>\n",
       "      <td>200219_BC_SA3_V19T26-031_CE_V1-Spot000001.h5ad</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>51712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumC1</td>\n",
       "      <td>V19T26-032_D1</td>\n",
       "      <td>200221_BC_SA3_C-D_V19T26-032_CE_V4-Spot000001....</td>\n",
       "      <td>d1.032</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumC2</td>\n",
       "      <td>V19T26-032_C1</td>\n",
       "      <td>200221_BC_SA3_C-D_V19T26-032_CE_V3-Spot000001....</td>\n",
       "      <td>c1.032</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumD1</td>\n",
       "      <td>V19T26-032_B1</td>\n",
       "      <td>200221_BC_SA3_C-D_V19T26-032_CE_V2-Spot000001....</td>\n",
       "      <td>b1.032</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumD2</td>\n",
       "      <td>V19T26-032_A1</td>\n",
       "      <td>200221_BC_SA3_C-D_V19T26-032_CE_V1-Spot000001....</td>\n",
       "      <td>a1.032</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumD3</td>\n",
       "      <td>V10F24-114_A1</td>\n",
       "      <td>200715_BC_SA2_D_E_V10F24-114.A1-Spot000001.h5ad</td>\n",
       "      <td>a1.114</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumD4</td>\n",
       "      <td>V10F24-114_B1</td>\n",
       "      <td>200715_BC_SA2_D_E_V10F24-114.B1-Spot000001.h5ad</td>\n",
       "      <td>b1.114</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>TD-2532-BCSA4</td>\n",
       "      <td>BCSA4</td>\n",
       "      <td>TumA1</td>\n",
       "      <td>V10B01-135_A1</td>\n",
       "      <td>220302_sATAC_OMNI_V10B01-135_NB.A1-Spot000001....</td>\n",
       "      <td>a1.135</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Patient Patientid Tumor area      Visium ID  \\\n",
       "0   TD-2532-BCSA1     BCSA1      TumA1  V19T26-012_A1   \n",
       "1   TD-2532-BCSA1     BCSA1      TumA2  V19T26-012_B1   \n",
       "2   TD-2532-BCSA1     BCSA1      TumB1  V19T26-012_C1   \n",
       "3   TD-2532-BCSA1     BCSA1      TumB2  V19T26-012_D1   \n",
       "4   TD-2532-BCSA2     BCSA2      TumB1  V10F24-112_C1   \n",
       "5   TD-2532-BCSA2     BCSA2      TumB2  V10F24-112_D1   \n",
       "6   TD-2532-BCSA2     BCSA2      TumC1  V10F24-113_A1   \n",
       "7   TD-2532-BCSA2     BCSA2      TumC2  V10F24-113_B1   \n",
       "8   TD-2532-BCSA2     BCSA2      TumD1  V10F24-113_C1   \n",
       "9   TD-2532-BCSA2     BCSA2      TumD2  V10F24-113_D1   \n",
       "10  TD-2532-BCSA2     BCSA2      TumE1  V10F24-114_C1   \n",
       "11  TD-2532-BCSA2     BCSA2      TumE2  V10F24-114_D1   \n",
       "12  TD-2532-BCSA3     BCSA3      TumA1  V19T26-031_D1   \n",
       "13  TD-2532-BCSA3     BCSA3      TumA2  V19T26-031_C1   \n",
       "14  TD-2532-BCSA3     BCSA3      TumB1  V19T26-031_B1   \n",
       "15  TD-2532-BCSA3     BCSA3      TumB2  V19T26-031_A1   \n",
       "16  TD-2532-BCSA3     BCSA3      TumC1  V19T26-032_D1   \n",
       "17  TD-2532-BCSA3     BCSA3      TumC2  V19T26-032_C1   \n",
       "18  TD-2532-BCSA3     BCSA3      TumD1  V19T26-032_B1   \n",
       "19  TD-2532-BCSA3     BCSA3      TumD2  V19T26-032_A1   \n",
       "20  TD-2532-BCSA2     BCSA2      TumD3  V10F24-114_A1   \n",
       "21  TD-2532-BCSA2     BCSA2      TumD4  V10F24-114_B1   \n",
       "22  TD-2532-BCSA4     BCSA4      TumA1  V10B01-135_A1   \n",
       "\n",
       "                                        Visum h5ad ID    name   type  width  \\\n",
       "0   200116_ST-AR_BC_A_B_V19T26-012_KT_V1-Spot00000...  a1.012   TNBC  29184   \n",
       "1   200116_ST-AR_BC_A_B_V19T26-012_KT_V2-Spot00000...  b1.012   TNBC  29184   \n",
       "2   200116_ST-AR_BC_A_B_V19T26-012_KT_V3-Spot00000...  c1.012   TNBC  29184   \n",
       "3   200116_ST-AR_BC_A_B_V19T26-012_KT_V4-Spot00000...  d1.012   TNBC  29184   \n",
       "4      200616_BC-SA2_V10F24-112_KT.C1-Spot000001.h5ad  c1.112  HER2+  47616   \n",
       "5      200616_BC-SA2_V10F24-112_KT.D1-Spot000001.h5ad  d1.112  HER2+  47616   \n",
       "6     200715_BC_SA2_C_D_V10F24-113.A1-Spot000001.h5ad  a1.113  HER2+  48640   \n",
       "7     200715_BC_SA2_C_D_V10F24-113.B1-Spot000001.h5ad  b1.113  HER2+  48640   \n",
       "8     200715_BC_SA2_C_D_V10F24-113.C1-Spot000001.h5ad  c1.113  HER2+  48640   \n",
       "9     200715_BC_SA2_C_D_V10F24-113.D1-Spot000001.h5ad  d1.113  HER2+  48640   \n",
       "10    200715_BC_SA2_D_E_V10F24-114.C1-Spot000001.h5ad  c1.114  HER2+  48640   \n",
       "11    200715_BC_SA2_D_E_V10F24-114.D1-Spot000001.h5ad  d1.114  HER2+  48640   \n",
       "12     200219_BC_SA3_V19T26-031_CE_V4-Spot000001.h5ad  d1.031  HER2+  52736   \n",
       "13     200219_BC_SA3_V19T26-031_CE_V3-Spot000001.h5ad  c1.031  HER2+  52736   \n",
       "14     200219_BC_SA3_V19T26-031_CE_V2-Spot000001.h5ad  b1.031  HER2+  52736   \n",
       "15     200219_BC_SA3_V19T26-031_CE_V1-Spot000001.h5ad     NaN  HER2+  52736   \n",
       "16  200221_BC_SA3_C-D_V19T26-032_CE_V4-Spot000001....  d1.032  HER2+  52736   \n",
       "17  200221_BC_SA3_C-D_V19T26-032_CE_V3-Spot000001....  c1.032  HER2+  52736   \n",
       "18  200221_BC_SA3_C-D_V19T26-032_CE_V2-Spot000001....  b1.032  HER2+  52736   \n",
       "19  200221_BC_SA3_C-D_V19T26-032_CE_V1-Spot000001....  a1.032  HER2+  52736   \n",
       "20    200715_BC_SA2_D_E_V10F24-114.A1-Spot000001.h5ad  a1.114  HER2+  48640   \n",
       "21    200715_BC_SA2_D_E_V10F24-114.B1-Spot000001.h5ad  b1.114  HER2+  48640   \n",
       "22  220302_sATAC_OMNI_V10B01-135_NB.A1-Spot000001....  a1.135  HER2+  48640   \n",
       "\n",
       "    height  \n",
       "0    29696  \n",
       "1    29696  \n",
       "2    29696  \n",
       "3    29696  \n",
       "4    48128  \n",
       "5    48128  \n",
       "6    47104  \n",
       "7    47104  \n",
       "8    47104  \n",
       "9    47104  \n",
       "10   47104  \n",
       "11   47104  \n",
       "12   52736  \n",
       "13   52736  \n",
       "14   52736  \n",
       "15   51712  \n",
       "16   52736  \n",
       "17   52736  \n",
       "18   52736  \n",
       "19   52736  \n",
       "20   47104  \n",
       "21   47104  \n",
       "22   47104  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load metadata\n",
    "info = pd.read_csv('/Users/tili/Desktop/CIIR/data/ST-AR_Visium_samples.csv')\n",
    "info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5833b140",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Patientid</th>\n",
       "      <th>Tumor area</th>\n",
       "      <th>Visium ID</th>\n",
       "      <th>Visum h5ad ID</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>width</th>\n",
       "      <th>height</th>\n",
       "      <th>sampleID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TD-2532-BCSA1</td>\n",
       "      <td>BCSA1</td>\n",
       "      <td>TumA1</td>\n",
       "      <td>V19T26-012_A1</td>\n",
       "      <td>200116_ST-AR_BC_A_B_V19T26-012_KT_V1-Spot00000...</td>\n",
       "      <td>a1.012</td>\n",
       "      <td>TNBC</td>\n",
       "      <td>29184</td>\n",
       "      <td>29696</td>\n",
       "      <td>BCSA1TumA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TD-2532-BCSA1</td>\n",
       "      <td>BCSA1</td>\n",
       "      <td>TumA2</td>\n",
       "      <td>V19T26-012_B1</td>\n",
       "      <td>200116_ST-AR_BC_A_B_V19T26-012_KT_V2-Spot00000...</td>\n",
       "      <td>b1.012</td>\n",
       "      <td>TNBC</td>\n",
       "      <td>29184</td>\n",
       "      <td>29696</td>\n",
       "      <td>BCSA1TumA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TD-2532-BCSA1</td>\n",
       "      <td>BCSA1</td>\n",
       "      <td>TumB1</td>\n",
       "      <td>V19T26-012_C1</td>\n",
       "      <td>200116_ST-AR_BC_A_B_V19T26-012_KT_V3-Spot00000...</td>\n",
       "      <td>c1.012</td>\n",
       "      <td>TNBC</td>\n",
       "      <td>29184</td>\n",
       "      <td>29696</td>\n",
       "      <td>BCSA1TumB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TD-2532-BCSA1</td>\n",
       "      <td>BCSA1</td>\n",
       "      <td>TumB2</td>\n",
       "      <td>V19T26-012_D1</td>\n",
       "      <td>200116_ST-AR_BC_A_B_V19T26-012_KT_V4-Spot00000...</td>\n",
       "      <td>d1.012</td>\n",
       "      <td>TNBC</td>\n",
       "      <td>29184</td>\n",
       "      <td>29696</td>\n",
       "      <td>BCSA1TumB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumB1</td>\n",
       "      <td>V10F24-112_C1</td>\n",
       "      <td>200616_BC-SA2_V10F24-112_KT.C1-Spot000001.h5ad</td>\n",
       "      <td>c1.112</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>47616</td>\n",
       "      <td>48128</td>\n",
       "      <td>BCSA2TumB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumB2</td>\n",
       "      <td>V10F24-112_D1</td>\n",
       "      <td>200616_BC-SA2_V10F24-112_KT.D1-Spot000001.h5ad</td>\n",
       "      <td>d1.112</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>47616</td>\n",
       "      <td>48128</td>\n",
       "      <td>BCSA2TumB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumC1</td>\n",
       "      <td>V10F24-113_A1</td>\n",
       "      <td>200715_BC_SA2_C_D_V10F24-113.A1-Spot000001.h5ad</td>\n",
       "      <td>a1.113</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "      <td>BCSA2TumC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumC2</td>\n",
       "      <td>V10F24-113_B1</td>\n",
       "      <td>200715_BC_SA2_C_D_V10F24-113.B1-Spot000001.h5ad</td>\n",
       "      <td>b1.113</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "      <td>BCSA2TumC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumD1</td>\n",
       "      <td>V10F24-113_C1</td>\n",
       "      <td>200715_BC_SA2_C_D_V10F24-113.C1-Spot000001.h5ad</td>\n",
       "      <td>c1.113</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "      <td>BCSA2TumD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumD2</td>\n",
       "      <td>V10F24-113_D1</td>\n",
       "      <td>200715_BC_SA2_C_D_V10F24-113.D1-Spot000001.h5ad</td>\n",
       "      <td>d1.113</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "      <td>BCSA2TumD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumE1</td>\n",
       "      <td>V10F24-114_C1</td>\n",
       "      <td>200715_BC_SA2_D_E_V10F24-114.C1-Spot000001.h5ad</td>\n",
       "      <td>c1.114</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "      <td>BCSA2TumE1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumE2</td>\n",
       "      <td>V10F24-114_D1</td>\n",
       "      <td>200715_BC_SA2_D_E_V10F24-114.D1-Spot000001.h5ad</td>\n",
       "      <td>d1.114</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "      <td>BCSA2TumE2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumA1</td>\n",
       "      <td>V19T26-031_D1</td>\n",
       "      <td>200219_BC_SA3_V19T26-031_CE_V4-Spot000001.h5ad</td>\n",
       "      <td>d1.031</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "      <td>BCSA3TumA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumA2</td>\n",
       "      <td>V19T26-031_C1</td>\n",
       "      <td>200219_BC_SA3_V19T26-031_CE_V3-Spot000001.h5ad</td>\n",
       "      <td>c1.031</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "      <td>BCSA3TumA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumB1</td>\n",
       "      <td>V19T26-031_B1</td>\n",
       "      <td>200219_BC_SA3_V19T26-031_CE_V2-Spot000001.h5ad</td>\n",
       "      <td>b1.031</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "      <td>BCSA3TumB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumB2</td>\n",
       "      <td>V19T26-031_A1</td>\n",
       "      <td>200219_BC_SA3_V19T26-031_CE_V1-Spot000001.h5ad</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>51712</td>\n",
       "      <td>BCSA3TumB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumC1</td>\n",
       "      <td>V19T26-032_D1</td>\n",
       "      <td>200221_BC_SA3_C-D_V19T26-032_CE_V4-Spot000001....</td>\n",
       "      <td>d1.032</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "      <td>BCSA3TumC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumC2</td>\n",
       "      <td>V19T26-032_C1</td>\n",
       "      <td>200221_BC_SA3_C-D_V19T26-032_CE_V3-Spot000001....</td>\n",
       "      <td>c1.032</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "      <td>BCSA3TumC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumD1</td>\n",
       "      <td>V19T26-032_B1</td>\n",
       "      <td>200221_BC_SA3_C-D_V19T26-032_CE_V2-Spot000001....</td>\n",
       "      <td>b1.032</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "      <td>BCSA3TumD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TD-2532-BCSA3</td>\n",
       "      <td>BCSA3</td>\n",
       "      <td>TumD2</td>\n",
       "      <td>V19T26-032_A1</td>\n",
       "      <td>200221_BC_SA3_C-D_V19T26-032_CE_V1-Spot000001....</td>\n",
       "      <td>a1.032</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>52736</td>\n",
       "      <td>52736</td>\n",
       "      <td>BCSA3TumD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumD3</td>\n",
       "      <td>V10F24-114_A1</td>\n",
       "      <td>200715_BC_SA2_D_E_V10F24-114.A1-Spot000001.h5ad</td>\n",
       "      <td>a1.114</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "      <td>BCSA2TumD3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>TD-2532-BCSA2</td>\n",
       "      <td>BCSA2</td>\n",
       "      <td>TumD4</td>\n",
       "      <td>V10F24-114_B1</td>\n",
       "      <td>200715_BC_SA2_D_E_V10F24-114.B1-Spot000001.h5ad</td>\n",
       "      <td>b1.114</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "      <td>BCSA2TumD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>TD-2532-BCSA4</td>\n",
       "      <td>BCSA4</td>\n",
       "      <td>TumA1</td>\n",
       "      <td>V10B01-135_A1</td>\n",
       "      <td>220302_sATAC_OMNI_V10B01-135_NB.A1-Spot000001....</td>\n",
       "      <td>a1.135</td>\n",
       "      <td>HER2+</td>\n",
       "      <td>48640</td>\n",
       "      <td>47104</td>\n",
       "      <td>BCSA4TumA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Patient Patientid Tumor area      Visium ID  \\\n",
       "0   TD-2532-BCSA1     BCSA1      TumA1  V19T26-012_A1   \n",
       "1   TD-2532-BCSA1     BCSA1      TumA2  V19T26-012_B1   \n",
       "2   TD-2532-BCSA1     BCSA1      TumB1  V19T26-012_C1   \n",
       "3   TD-2532-BCSA1     BCSA1      TumB2  V19T26-012_D1   \n",
       "4   TD-2532-BCSA2     BCSA2      TumB1  V10F24-112_C1   \n",
       "5   TD-2532-BCSA2     BCSA2      TumB2  V10F24-112_D1   \n",
       "6   TD-2532-BCSA2     BCSA2      TumC1  V10F24-113_A1   \n",
       "7   TD-2532-BCSA2     BCSA2      TumC2  V10F24-113_B1   \n",
       "8   TD-2532-BCSA2     BCSA2      TumD1  V10F24-113_C1   \n",
       "9   TD-2532-BCSA2     BCSA2      TumD2  V10F24-113_D1   \n",
       "10  TD-2532-BCSA2     BCSA2      TumE1  V10F24-114_C1   \n",
       "11  TD-2532-BCSA2     BCSA2      TumE2  V10F24-114_D1   \n",
       "12  TD-2532-BCSA3     BCSA3      TumA1  V19T26-031_D1   \n",
       "13  TD-2532-BCSA3     BCSA3      TumA2  V19T26-031_C1   \n",
       "14  TD-2532-BCSA3     BCSA3      TumB1  V19T26-031_B1   \n",
       "15  TD-2532-BCSA3     BCSA3      TumB2  V19T26-031_A1   \n",
       "16  TD-2532-BCSA3     BCSA3      TumC1  V19T26-032_D1   \n",
       "17  TD-2532-BCSA3     BCSA3      TumC2  V19T26-032_C1   \n",
       "18  TD-2532-BCSA3     BCSA3      TumD1  V19T26-032_B1   \n",
       "19  TD-2532-BCSA3     BCSA3      TumD2  V19T26-032_A1   \n",
       "20  TD-2532-BCSA2     BCSA2      TumD3  V10F24-114_A1   \n",
       "21  TD-2532-BCSA2     BCSA2      TumD4  V10F24-114_B1   \n",
       "22  TD-2532-BCSA4     BCSA4      TumA1  V10B01-135_A1   \n",
       "\n",
       "                                        Visum h5ad ID    name   type  width  \\\n",
       "0   200116_ST-AR_BC_A_B_V19T26-012_KT_V1-Spot00000...  a1.012   TNBC  29184   \n",
       "1   200116_ST-AR_BC_A_B_V19T26-012_KT_V2-Spot00000...  b1.012   TNBC  29184   \n",
       "2   200116_ST-AR_BC_A_B_V19T26-012_KT_V3-Spot00000...  c1.012   TNBC  29184   \n",
       "3   200116_ST-AR_BC_A_B_V19T26-012_KT_V4-Spot00000...  d1.012   TNBC  29184   \n",
       "4      200616_BC-SA2_V10F24-112_KT.C1-Spot000001.h5ad  c1.112  HER2+  47616   \n",
       "5      200616_BC-SA2_V10F24-112_KT.D1-Spot000001.h5ad  d1.112  HER2+  47616   \n",
       "6     200715_BC_SA2_C_D_V10F24-113.A1-Spot000001.h5ad  a1.113  HER2+  48640   \n",
       "7     200715_BC_SA2_C_D_V10F24-113.B1-Spot000001.h5ad  b1.113  HER2+  48640   \n",
       "8     200715_BC_SA2_C_D_V10F24-113.C1-Spot000001.h5ad  c1.113  HER2+  48640   \n",
       "9     200715_BC_SA2_C_D_V10F24-113.D1-Spot000001.h5ad  d1.113  HER2+  48640   \n",
       "10    200715_BC_SA2_D_E_V10F24-114.C1-Spot000001.h5ad  c1.114  HER2+  48640   \n",
       "11    200715_BC_SA2_D_E_V10F24-114.D1-Spot000001.h5ad  d1.114  HER2+  48640   \n",
       "12     200219_BC_SA3_V19T26-031_CE_V4-Spot000001.h5ad  d1.031  HER2+  52736   \n",
       "13     200219_BC_SA3_V19T26-031_CE_V3-Spot000001.h5ad  c1.031  HER2+  52736   \n",
       "14     200219_BC_SA3_V19T26-031_CE_V2-Spot000001.h5ad  b1.031  HER2+  52736   \n",
       "15     200219_BC_SA3_V19T26-031_CE_V1-Spot000001.h5ad     NaN  HER2+  52736   \n",
       "16  200221_BC_SA3_C-D_V19T26-032_CE_V4-Spot000001....  d1.032  HER2+  52736   \n",
       "17  200221_BC_SA3_C-D_V19T26-032_CE_V3-Spot000001....  c1.032  HER2+  52736   \n",
       "18  200221_BC_SA3_C-D_V19T26-032_CE_V2-Spot000001....  b1.032  HER2+  52736   \n",
       "19  200221_BC_SA3_C-D_V19T26-032_CE_V1-Spot000001....  a1.032  HER2+  52736   \n",
       "20    200715_BC_SA2_D_E_V10F24-114.A1-Spot000001.h5ad  a1.114  HER2+  48640   \n",
       "21    200715_BC_SA2_D_E_V10F24-114.B1-Spot000001.h5ad  b1.114  HER2+  48640   \n",
       "22  220302_sATAC_OMNI_V10B01-135_NB.A1-Spot000001....  a1.135  HER2+  48640   \n",
       "\n",
       "    height    sampleID  \n",
       "0    29696  BCSA1TumA1  \n",
       "1    29696  BCSA1TumA2  \n",
       "2    29696  BCSA1TumB1  \n",
       "3    29696  BCSA1TumB2  \n",
       "4    48128  BCSA2TumB1  \n",
       "5    48128  BCSA2TumB2  \n",
       "6    47104  BCSA2TumC1  \n",
       "7    47104  BCSA2TumC2  \n",
       "8    47104  BCSA2TumD1  \n",
       "9    47104  BCSA2TumD2  \n",
       "10   47104  BCSA2TumE1  \n",
       "11   47104  BCSA2TumE2  \n",
       "12   52736  BCSA3TumA1  \n",
       "13   52736  BCSA3TumA2  \n",
       "14   52736  BCSA3TumB1  \n",
       "15   51712  BCSA3TumB2  \n",
       "16   52736  BCSA3TumC1  \n",
       "17   52736  BCSA3TumC2  \n",
       "18   52736  BCSA3TumD1  \n",
       "19   52736  BCSA3TumD2  \n",
       "20   47104  BCSA2TumD3  \n",
       "21   47104  BCSA2TumD4  \n",
       "22   47104  BCSA4TumA1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get sample ID\n",
    "info['sampleID'] = info['Patientid'] + info['Tumor area']\n",
    "info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b0dd1dfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5539\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:5262 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:15903 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 5262 genes and rna_count_based density_prior in cells mode...\n",
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.363, KL reg: 0.294\n",
      "Score: 0.546, KL reg: 0.000\n",
      "Score: 0.560, KL reg: 0.000\n",
      "Score: 0.563, KL reg: 0.000\n",
      "Score: 0.564, KL reg: 0.000\n",
      "Score: 0.564, KL reg: 0.000\n",
      "Score: 0.564, KL reg: 0.000\n",
      "Score: 0.564, KL reg: 0.000\n",
      "Score: 0.565, KL reg: 0.000\n",
      "Score: 0.565, KL reg: 0.000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Saving results..\n",
      "INFO:root:spatial prediction dataframe is saved in `obsm` `tangram_ct_pred` of the spatial AnnData.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5539\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:5285 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:15829 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 5285 genes and rna_count_based density_prior in cells mode...\n",
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.298, KL reg: 0.308\n",
      "Score: 0.503, KL reg: 0.000\n",
      "Score: 0.518, KL reg: 0.000\n",
      "Score: 0.521, KL reg: 0.000\n",
      "Score: 0.522, KL reg: 0.000\n",
      "Score: 0.523, KL reg: 0.000\n",
      "Score: 0.523, KL reg: 0.000\n",
      "Score: 0.523, KL reg: 0.000\n",
      "Score: 0.523, KL reg: 0.000\n",
      "Score: 0.523, KL reg: 0.000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Saving results..\n",
      "INFO:root:spatial prediction dataframe is saved in `obsm` `tangram_ct_pred` of the spatial AnnData.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5539\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:5262 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:15631 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 5262 genes and rna_count_based density_prior in cells mode...\n",
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.292, KL reg: 0.341\n",
      "Score: 0.502, KL reg: 0.000\n",
      "Score: 0.517, KL reg: 0.000\n",
      "Score: 0.520, KL reg: 0.000\n",
      "Score: 0.521, KL reg: 0.000\n",
      "Score: 0.521, KL reg: 0.000\n",
      "Score: 0.522, KL reg: 0.000\n",
      "Score: 0.522, KL reg: 0.000\n",
      "Score: 0.522, KL reg: 0.000\n",
      "Score: 0.522, KL reg: 0.000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Saving results..\n",
      "INFO:root:spatial prediction dataframe is saved in `obsm` `tangram_ct_pred` of the spatial AnnData.\n"
     ]
    }
   ],
   "source": [
    "# Map the ST data using marker genes. All datasets needs to be unzipped, \n",
    "# resulting h5ad and csv files should be placed in the data folder\n",
    "\n",
    "for row in info.iterrows():\n",
    "    id = row[1]['sampleID']\n",
    "    # Load ST data\n",
    "    path = '/Users/tili/Desktop/CIIR/data/ST_data/Tangram/'+id+\".h5ad\"\n",
    "    ad_sp = sc.read_h5ad(path)\n",
    "    if row[1]['type'] == \"TNBC\":\n",
    "        path = os.path.join('/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/Tangram/','tnbc.h5ad')\n",
    "        ad_sc = sc.read_h5ad(path)\n",
    "    elif row[1]['type'] == \"HER2+\":\n",
    "        path = os.path.join('/Users/tili/Desktop/CIIR/data/Wu_etal_2021_BRCA_scRNASeq/Tangram/','her2.h5ad')\n",
    "        ad_sc = sc.read_h5ad(path)\n",
    "    \n",
    "    # Usually, the data input should be in raw count, especially if the ST data are in raw count form\n",
    "    np.unique(ad_sc.X.toarray()[0, :])\n",
    "\n",
    "    # Here, we only do some light pre-processing as library size correction (in scanpy, via sc.pp.normalize) \n",
    "    # to normalize the number of count within each cell to a fixed number.\n",
    "\n",
    "    # Sometimes, we apply more sophisticated pre-processing methods, for example for batch correction, \n",
    "    # although mapping works great with raw data.\n",
    "\n",
    "    # Ideally, the single cell and spatial datasets, should exhibit signals as similar as possible and \n",
    "    # the pre-processing pipeline should be finalized to harmonize the signals\n",
    "    #sc.pp.normalize_total(ad_sc)\n",
    "\n",
    "    # Cell types are annotated in the subclass_label field, for which we plot cell counts.\n",
    "    # Note that cell type proportion should be similar in the two datasets\n",
    "    ad_sc.obs.celltype_minor.value_counts()\n",
    "    \n",
    "    # Tangram learns a spatial alignment of the single cell data so that the gene expression of the \n",
    "    # aligned single cell data is as similar as possible to that of the spatial data\n",
    "\n",
    "    # Tangram only looks at a subset genes, specified by the user, called the training genes.\n",
    "    # they need to bear interesting signals and to be measured with high quality.\n",
    "\n",
    "    # Typically, a good start is to choose 100-1000 top marker genes, evenly stratified across cell types. \n",
    "    # Sometimes, we also use the entire transcriptome, \n",
    "    # or perform different mappings using different set of training genes to see how much the result change.\n",
    "\n",
    "    df_genes = pd.read_csv('/Users/tili/Desktop/CIIR/data/ST_data/Tangram/features.csv', index_col=0)\n",
    "    markers = np.reshape(df_genes.values, (-1, ))\n",
    "    markers = list(markers)\n",
    "    print(len(markers))\n",
    "    \n",
    "    # We now need to prepare the datasets for mapping by creating training_genes \n",
    "    # field in uns dictionary of the two AnnData structures.\n",
    "\n",
    "    # This training_genes field contains genes subset on the list of training genes. \n",
    "    # This field will be used later inside the mapping function to create training datasets.\n",
    "\n",
    "    # Also, the gene order needs to be the same in the datasets. This is because Tangram maps \n",
    "    # using only gene expression, so the j-th column in each matrix must correspond to the same gene.\n",
    "\n",
    "    # And if data entries of a gene are all zero, this gene will be removed.\n",
    "\n",
    "    # In the pp_adatas function, the gene names are converted to lower case to get rid of the inconsistent capitalization. \n",
    "    # If this is not wanted, you can set the parameter gene_to_lowercase = False\n",
    "    tg.pp_adatas(ad_sc, ad_sp, genes=markers)\n",
    "    \n",
    "    # the marker genes need to be shared by both dataset. If a gene is missing, pp_adatas will just take it out.\n",
    "    # Finally, the assert line below is a good way to ensure that the genes in the training_genes field in uns \n",
    "    # are actually ordered in both AnnDatas.\n",
    "    assert ad_sc.uns['training_genes'] == ad_sp.uns['training_genes']\n",
    "    \n",
    "    # Mapping should be interrupted after the score plateaus,\n",
    "    # which can be controlled by passing the num_epochs parameter.\n",
    "\n",
    "    # The score measures the similarity between the gene expression of the mapped cells \n",
    "    # vs spatial data: higher score means better mapping.\n",
    "\n",
    "    # Note that we obtained excellent mapping even if Tangram converges to a low scores (the typical case is \n",
    "    # when the spatial data are very sparse): we use the score merely to assess convergence\n",
    "\n",
    "    # If you are running Tangram with a GPU, uncomment device=cuda:0 and comment the line device=cpu.\n",
    "\n",
    "    # For this basic mapping, we do not use regularizers. \n",
    "    # More sophisticated loss functions can be used using the Tangram library.\n",
    "    # For example, you can pass your density_prior with the hyperparameter lambda_d to regularize the spatial \n",
    "    # density of cells. Currently uniform, rna_count_based and \n",
    "    # customized input array are supported for density_prior argument.\n",
    "\n",
    "    # Instead of mapping single cells, we can \"average\" the cells within a cluster and map the averaged cells instead, \n",
    "    # which drammatically improves performances. This suggestion was proposed by Sten Linnarsson. \n",
    "    # To activate this mode, select mode='clusters' and pass the annotation field to cluster_label.\n",
    "    ad_map = tg.map_cells_to_space(\n",
    "        adata_sc=ad_sc,\n",
    "        adata_sp=ad_sp,\n",
    "        device='cpu',\n",
    "        #mode='cells', # defualt mode \n",
    "        #density_prior='rna_count_based', #default method\n",
    "        #device='cuda:0',\n",
    "    )\n",
    "    \n",
    "    # The mapping results are stored in the returned AnnData structure, saved as ad_map, structured as following:\n",
    "    # The cell-by-spot matrix X contains the probability of cell i to be in spot j.\n",
    "    # The obs dataframe contains the metadata of the single cells.\n",
    "    # The var dataframe contains the metadata of the spatial data.\n",
    "    # The uns dictionary contains a dataframe with various information about the training genes \n",
    "    # (saved ad train_genes_df).\n",
    "\n",
    "    # The most common application for mapping single cell data onto space is to transfer the cell \n",
    "    # type annotations onto space, dona via plot_cell_annotation, which visualizes spatial probability maps \n",
    "    # of the annotation in the obs dataframe (here, the subclass_label field). You can set robust argument to \n",
    "    # True and at the same time set the perc argument to set the range to the colormap, which would help remove outliers.\n",
    "\n",
    "    ad_map\n",
    "    tg.project_cell_annotations(ad_map, ad_sp, annotation='celltype_minor')\n",
    "\n",
    "    # Save the results\n",
    "    path = \"/Users/tili/Desktop/CIIR/results/Tangram/Minor_HER2_genes/\" + id + \"_Tangram.csv\"\n",
    "    ad_sp.obsm[\"tangram_ct_pred\"].to_csv(path, index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7af77460",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
